NuRadioMC/EvtGen/generator.py

```diff
@@ -460,9 +371,134 @@ def J(E):
     return energies
 
 
+def generate_vertex_positions(volume, proposal, attributes):
+    """
+    helper function that generates the vertex position randomly distributed in simulation volume. 
+    The relevant quantities are also saved into the hdf5 attributes
+    
+    Parameters
+    volume: dictionarty
+        dict specifying the volume
+    proposal: bool
+        specifies if secondary interaction via proposal are calculated
+    attributes: dicitionary
+        dict storing hdf5 attributes
+    """
+    n_events = attributes['n_events']
+    if("fiducial_rmax" in volume):  # user specifies a cylinder
+        if('fiducial_rmin' in volume):
+            attributes['fiducial_rmin'] = volume['fiducial_rmin']
+        else:
+            attributes['fiducial_rmin'] = 0
+        attributes['fiducial_rmax'] = volume['fiducial_rmax']
+        attributes['fiducial_zmin'] = volume['fiducial_zmin']
+        attributes['fiducial_zmax'] = volume['fiducial_zmax']
+
+        rmin = attributes['fiducial_rmin']
+        rmax = attributes['fiducial_rmax']
+        zmin = attributes['fiducial_zmin']
+        zmax = attributes['fiducial_zmax']
+
+        # We increase the radius of the cylinder according to the tau track length
+        if(proposal):
+            tau_95_length = get_tau_95_length(attributes['Emax'])
+            if("full_rmin" in volume):
+                rmin = volume['full_rmin']
+            else:
+                rmin = attributes['fiducial_rmin'] / 3.
+            if('full_rmax' in volume):
+                rmax = volume['full_rmax']
+            else:
+                full_rmax = tau_95_length + attributes['fiducial_rmax']
+            if('full_zmax' in volume):
+                zmax = volume['full_zmax']
+            else:
+                zmax = attributes['fiducial_zmax'] / 3.
+            if('full_zmin' in volume):
+                zmin = volume['full_zmin']
+            else:
+                zmin = attributes['fiducial_zmin'] - tau_95_length  # we have a minus sign here because the zmin coordinate is negative
+            # increase the total number of events such that we end up with the same number of events in the fiducial volume
+            n_events = n_events * int((rmax / attributes['fiducial_rmax']) ** 2 * zmin / attributes['fiducial_zmin'])
```

Comment:
Shouldn't we account for the vertical length differently here?

Reply:
namely, zmin/fiducial_zmin is not the same as (zmax-zmin)/(fiducial_zmax-fiducial_zmin)

Reply:
good catch. I changed it.